File: src/core/tax.ts
/**
 * Steuer-Engine: Kleinunternehmer (§19 UStG) vs. Regelbesteuerung.
 * Portierung von src-tauri/src/tax.rs.
 *
 * Konventionen: Geld in Cent, Steuersatz in Basispunkten (1900 = 19 %),
 * Menge in Tausendstel (1000 = 1,000 Einheiten). Prozentuale Zu-/Abschläge
 * werden ebenfalls in Basispunkten geführt (3000 = 30 %).
 */

export const KLEINUNTERNEHMER_HINWEIS =
  "Gemäß § 19 UStG wird keine Umsatzsteuer berechnet.";

/** Erfassungsart eines Zu-/Abschlags: prozentual (Wert in bp) oder absolut (Cent). */
export type AdjustmentType = "percent" | "amount";

/** Ein Zu- oder Abschlag (Rabatt/Aufpreis). */
export interface Adjustment {
  type: AdjustmentType;
  /** percent → Basispunkte (3000 = 30 %); amount → Cent. */
  value: number;
  reason?: string | null;
}

export interface LineInput {
  quantityMilli: number;
  unitPriceNetCents: number;
  taxRateBp: number;
  /** Positions-Rabatt (EN 16931 BG-27, Allowance). */
  discount?: Adjustment | null;
  /** Positions-Aufpreis (EN 16931 BG-28, Charge). */
  surcharge?: Adjustment | null;
}

export interface LineResult {
  effectiveTaxRateBp: number;
  /** Positionsbasis = Menge × Einzelpreis, vor Zu-/Abschlag. */
  baseCents: number;
  /** Rabattbetrag (Allowance) dieser Position. */
  discountCents: number;
  /** Aufpreisbetrag (Charge) dieser Position. */
  surchargeCents: number;
  /** Positions-Netto nach Zu-/Abschlag (EN 16931 BT-131). */
  netCents: number;
}

export interface TaxBreakdownRow {
  taxRateBp: number;
  /** Steuerbasis je Satz nach Rechnungs-Rabatt (EN 16931 BT-116). */
  netCents: number;
  taxCents: number;
}

/** Aufteilung des rechnungsweiten Rabatts je Steuersatz (EN 16931 BG-20). */
export interface AllowanceByRate {
  taxRateBp: number;
  amountCents: number;
}

export interface InvoiceTotals {
  isKleinunternehmer: boolean;
  lines: LineResult[];
  breakdown: TaxBreakdownRow[];
  /** Summe der Positions-Nettobeträge vor Rechnungs-Rabatt (EN 16931 BT-106). */
  lineNetSumCents: number;
  /** Rechnungsweiter Rabatt gesamt (EN 16931 BT-107). */
  invoiceDiscountCents: number;
  /** Aufteilung des Rechnungs-Rabatts je Steuersatz (EN 16931 BG-20). */
  allowancesByRate: AllowanceByRate[];
  /** Steuerbasis gesamt = BT-106 − BT-107 (EN 16931 BT-109). */
  netTotalCents: number;
  taxTotalCents: number;
  grossTotalCents: number;
  legalNote: string | null;
}

/** Kaufmännische Rundung (round half away from zero), ganzzahlig. */
export function roundDiv(numerator: number, denominator: number): number {
  const half = Math.trunc(denominator / 2);
  return numerator >= 0
    ? Math.trunc((numerator + half) / denominator)
    : Math.trunc((numerator - half) / denominator);
}

/** Betrag eines Zu-/Abschlags bezogen auf eine Basis (in Cent). */
export function adjustmentAmount(
  adj: Adjustment | null | undefined,
  baseCents: number,
): number {
  if (!adj) return 0;
  return adj.type === "percent" ? roundDiv(baseCents * adj.value, 10_000) : adj.value;
}

/**
 * USt wird gemäß EN 16931 je Steuersatz auf die Netto-Summe berechnet
 * (nicht je Zeile) – vermeidet Rundungsdifferenzen zum Steuerausweis.
 *
 * Positions-Zu-/Abschläge fließen in das Positions-Netto (BT-131) ein, ein
 * rechnungsweiter Rabatt (BG-20) wird anteilig je Steuersatz auf die
 * Steuerbasis angerechnet – so bleibt der Ausweis EN-16931-konsistent.
 */
export function computeInvoiceTotals(
  lines: LineInput[],
  isKleinunternehmer: boolean,
  invoiceDiscount?: Adjustment | null,
): InvoiceTotals {
  const lineResults: LineResult[] = [];
  const netByRate = new Map<number, number>();

  for (const line of lines) {
    const baseCents = roundDiv(line.quantityMilli * line.unitPriceNetCents, 1000);
    const rate = isKleinunternehmer ? 0 : line.taxRateBp;
    const discountCents = adjustmentAmount(line.discount, baseCents);
    const surchargeCents = adjustmentAmount(line.surcharge, baseCents);
    const netCents = baseCents - discountCents + surchargeCents;
    netByRate.set(rate, (netByRate.get(rate) ?? 0) + netCents);
    lineResults.push({ effectiveTaxRateBp: rate, baseCents, discountCents, surchargeCents, netCents });
  }

  const lineNetSumCents = [...netByRate.values()].reduce((a, b) => a + b, 0);

  // Rechnungsweiter Rabatt (BG-20) anteilig je Steuersatz aufteilen, damit die
  // Steuerbasis je Satz stimmt. Die Rundungsdifferenz landet auf dem
  // umsatzstärksten Satz, sodass die Summe der Anteile exakt dem Rabatt entspricht.
  let invoiceDiscountCents = 0;
  const allowancesByRate: AllowanceByRate[] = [];
  if (invoiceDiscount && lineNetSumCents > 0) {
    const total = Math.min(
      adjustmentAmount(invoiceDiscount, lineNetSumCents),
      lineNetSumCents,
    );
    if (total > 0) {
      invoiceDiscountCents = total;
      const rates = [...netByRate.keys()].sort((a, b) => a - b);
      const alloc = new Map<number, number>();
      let allocated = 0;
      for (const rate of rates) {
        const a = roundDiv(total * netByRate.get(rate)!, lineNetSumCents);
        alloc.set(rate, a);
        allocated += a;
      }
      const diff = total - allocated;
      if (diff !== 0) {
        const biggest = rates.reduce(
          (m, r) => (netByRate.get(r)! > netByRate.get(m)! ? r : m),
          rates[0],
        );
        alloc.set(biggest, alloc.get(biggest)! + diff);
      }
      for (const rate of rates) {
        const a = alloc.get(rate)!;
        netByRate.set(rate, netByRate.get(rate)! - a);
        if (a !== 0) allowancesByRate.push({ taxRateBp: rate, amountCents: a });
      }
    }
  }

  const breakdown: TaxBreakdownRow[] = [];
  let netTotal = 0;
  let taxTotal = 0;
  for (const rateBp of [...netByRate.keys()].sort((a, b) => a - b)) {
    const netCents = netByRate.get(rateBp)!;
    const taxCents = roundDiv(netCents * rateBp, 10_000);
    breakdown.push({ taxRateBp: rateBp, netCents, taxCents });
    netTotal += netCents;
    taxTotal += taxCents;
  }

  return {
    isKleinunternehmer,
    lines: lineResults,
    breakdown,
    lineNetSumCents,
    invoiceDiscountCents,
    allowancesByRate,
    netTotalCents: netTotal,
    taxTotalCents: taxTotal,
    grossTotalCents: netTotal + taxTotal,
    legalNote: isKleinunternehmer ? KLEINUNTERNEHMER_HINWEIS : null,
  };
}

File: src-tauri/src/tax.rs
//! Steuer-Engine: Kleinunternehmer (§19 UStG) vs. Regelbesteuerung.
//!
//! Konventionen (siehe auch db/migrations/0001_init.sql):
//!   * Geldbeträge in Cent (i64).
//!   * Steuersätze in Basispunkten (bp): 19,00 % = 1900.
//!   * Mengen in Tausendstel (milli): 1,000 Einheiten = 1000.
//!
//! Reine Logik ohne DB-Abhängigkeit -> vollständig unit-testbar.

use std::collections::BTreeMap;

/// Pflichthinweis für Kleinunternehmer nach §19 UStG.
pub const KLEINUNTERNEHMER_HINWEIS: &str =
    "Gemäß § 19 UStG wird keine Umsatzsteuer berechnet.";

#[derive(Clone, Copy, Debug)]
pub struct TaxContext {
    pub is_kleinunternehmer: bool,
}

#[derive(Clone, Copy, Debug)]
pub struct LineInput {
    pub quantity_milli: i64,
    pub unit_price_net_cents: i64,
    /// Wird im KU-Modus auf 0 gezwungen.
    pub tax_rate_bp: i32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct LineResult {
    pub effective_tax_rate_bp: i32,
    pub net_cents: i64,
}

/// Steueraufschlüsselung je Satz – Basis für Rechnungsausweis und UStVA.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TaxBreakdownRow {
    pub tax_rate_bp: i32,
    pub net_cents: i64,
    pub tax_cents: i64,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct InvoiceTotals {
    pub is_kleinunternehmer: bool,
    pub lines: Vec<LineResult>,
    pub breakdown: Vec<TaxBreakdownRow>,
    pub net_total_cents: i64,
    pub tax_total_cents: i64,
    pub gross_total_cents: i64,
    pub legal_note: Option<String>,
}

/// Kaufmännische Rundung (round half away from zero). i128 gegen Überlauf.
fn round_div(numerator: i128, denominator: i128) -> i64 {
    debug_assert!(denominator > 0);
    let half = denominator / 2;
    let result = if numerator >= 0 {
        (numerator + half) / denominator
    } else {
        (numerator - half) / denominator
    };
    result as i64
}

pub fn compute_line(line: &LineInput, ctx: &TaxContext) -> LineResult {
    let net = round_div(
        line.quantity_milli as i128 * line.unit_price_net_cents as i128,
        1000,
    );
    let rate = if ctx.is_kleinunternehmer { 0 } else { line.tax_rate_bp };
    LineResult { effective_tax_rate_bp: rate, net_cents: net }
}

/// Berechnet Summen und Steueraufschlüsselung einer Rechnung.
///
/// Die USt wird gemäß EN 16931 je Steuersatz auf die Netto-Summe berechnet
/// (nicht je Zeile) – das vermeidet Rundungsdifferenzen gegenüber dem
/// separaten Steuerausweis auf der Rechnung.
pub fn compute_invoice_totals(lines: &[LineInput], ctx: &TaxContext) -> InvoiceTotals {
    let mut line_results = Vec::with_capacity(lines.len());
    let mut net_by_rate: BTreeMap<i32, i64> = BTreeMap::new();

    for line in lines {
        let res = compute_line(line, ctx);
        *net_by_rate.entry(res.effective_tax_rate_bp).or_default() += res.net_cents;
        line_results.push(res);
    }

    let mut breakdown = Vec::new();
    let (mut net_total, mut tax_total) = (0i64, 0i64);
    for (&rate_bp, &net) in &net_by_rate {
        let tax = round_div(net as i128 * rate_bp as i128, 10_000);
        breakdown.push(TaxBreakdownRow { tax_rate_bp: rate_bp, net_cents: net, tax_cents: tax });
        net_total += net;
        tax_total += tax;
    }

    let legal_note = if ctx.is_kleinunternehmer {
        Some(KLEINUNTERNEHMER_HINWEIS.to_string())
    } else {
        None
    };

    InvoiceTotals {
        is_kleinunternehmer: ctx.is_kleinunternehmer,
        lines: line_results,
        breakdown,
        net_total_cents: net_total,
        tax_total_cents: tax_total,
        gross_total_cents: net_total + tax_total,
        legal_note,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn line(qty_milli: i64, price: i64, rate: i32) -> LineInput {
        LineInput { quantity_milli: qty_milli, unit_price_net_cents: price, tax_rate_bp: rate }
    }

    #[test]
    fn regelbesteuerung_weist_ust_getrennt_aus() {
        let ctx = TaxContext { is_kleinunternehmer: false };
        // 2 x 100,00 € @ 19 %  +  1 x 50,00 € @ 7 %
        let lines = [line(2000, 10_000, 1900), line(1000, 5_000, 700)];
        let t = compute_invoice_totals(&lines, &ctx);

        assert_eq!(t.net_total_cents, 25_000);
        assert_eq!(t.tax_total_cents, 3_800 + 350); // 38,00 + 3,50
        assert_eq!(t.gross_total_cents, 29_150);
        assert_eq!(t.breakdown.len(), 2);
        assert!(t.legal_note.is_none());
    }

    #[test]
    fn kleinunternehmer_erzwingt_null_prozent_und_hinweis() {
        let ctx = TaxContext { is_kleinunternehmer: true };
        // Satz 19 wird ignoriert -> 0 %
        let lines = [line(1000, 10_000, 1900)];
        let t = compute_invoice_totals(&lines, &ctx);

        assert_eq!(t.tax_total_cents, 0);
        assert_eq!(t.gross_total_cents, 10_000);
        assert_eq!(t.breakdown[0].tax_rate_bp, 0);
        assert_eq!(t.legal_note.as_deref(), Some(KLEINUNTERNEHMER_HINWEIS));
    }
}
